### ISSUE TITLE
mro + undef interaction: S_unshare_hek_or_pvn: Assertion error

### ISSUE THREAD

**Issue by atoomic** [rurban/perl-compiler#255]
mro + undef interaction

S_unshare_hek_or_pvn: Assertion `he->shared_he_he.hent_hek == hek' failed.

````
 > echo 'require mro; mro::get_mro("Abc"); *Foo::ISA = *Bar::ISA; undef *UNIVERSAL::DESTROY; print qq/ok\n/' > test.pl
 > perl test.pl
ok
> perlcc -r test.pl # same error with -O3
pccQi1de.c:9355: warning: this decimal constant is unsigned only in ISO C90
pccQi1de.c:9355: warning: this decimal constant is unsigned only in ISO C90
pccQi1de.c:9356: warning: this decimal constant is unsigned only in ISO C90
pccQi1de.c:9356: warning: this decimal constant is unsigned only in ISO C90
pccQi1de.c:9357: warning: this decimal constant is unsigned only in ISO C90
pccQi1de.c:9357: warning: this decimal constant is unsigned only in ISO C90
pccQi1de.c:9937: warning: integer constant is too large for ‘unsigned long’ type
pccQi1de.c:9938: warning: integer constant is too large for ‘unsigned long’ type
test: hv.c:2784: S_unshare_hek_or_pvn: Assertion `he->shared_he_he.hent_hek == hek' failed.
````

but strangely the same code does not raise the exception when run as a oneliner
````
> perlcc -r -e "$(cat test.pl)"
pccpUI8j.c:9355: warning: this decimal constant is unsigned only in ISO C90
pccpUI8j.c:9355: warning: this decimal constant is unsigned only in ISO C90
pccpUI8j.c:9356: warning: this decimal constant is unsigned only in ISO C90
pccpUI8j.c:9356: warning: this decimal constant is unsigned only in ISO C90
pccpUI8j.c:9357: warning: this decimal constant is unsigned only in ISO C90
pccpUI8j.c:9357: warning: this decimal constant is unsigned only in ISO C90
pccpUI8j.c:9937: warning: integer constant is too large for ‘unsigned long’ type
pccpUI8j.c:9938: warning: integer constant is too large for ‘unsigned long’ type
ok
````

**Comment by toddr** [comment]
How do the generated C files differ?

**Comment by rurban** [comment]
That's out typical glob_assign_glob core problem, which can only be fixed in core, not the compiler.
I cannot repro the first case.

**Comment by toddr** [comment]
So this can be simplified to this and fails.

perlcc -r 'require mro; mro::get_mro("Abc"); print "ok\n";'

But this succeeds

perlcc -r 'BEGIN { require mro } mro::get_mro("Abc"); print "ok\n";'

**Comment by toddr** [comment]
BTW: -O3 fixes this.

**Comment by toddr** [comment]
This looks to be some sort of failure to compile mro in. I'll need to look at it tomorrow.

**Comment by rurban** [comment]
Please do not create .pl files in root, put them into t/ like t/issue255.pl
can repro it with the file without -v5. but with -v5 it works ok.

**Comment by rurban** [comment]
Testing a general fix for those assertions. I think I've found it
```
@@ -5875,7 +5875,7 @@ my_share_hek( pTHX_ const char *str, I32 len, register U32 hash ) {
     if (!hash) {
       PERL_HASH(hash, str, abs(len));
     }
-    return Perl_share_hek(aTHX_ str, len, hash);
+    return share_hek_hek(Perl_share_hek(aTHX_ str, len, hash));
 }
```

**Comment by atoomic** [comment]
fixed with #251 and commit 1fa99b0
